# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmValue.cpp
# Commit: 72bf9c871ded
# Full Hash: 72bf9c871ded006f56f09b24d0723365516445bb
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 5
# Description:
#   Bug 1692065 - wasm: Add tagging for 31-bit integers and i31ref. r=yury
#   
#   This commit adds a tag 0x1 for 31-bit integers in AnyRef. Any JS value
#   that is losslessly representable (without coercion) in this range is
#   stored without boxing. This is chosen to line up exactly with how the
# ==============================================================================

diff -r 28feda14fdec -r 72bf9c871ded js/src/wasm/WasmValue.cpp
--- a/js/src/wasm/WasmValue.cpp	Fri Aug 04 14:06:50 2023 +0000
+++ b/js/src/wasm/WasmValue.cpp	Fri Aug 04 14:06:50 2023 +0000
@@ -138,6 +138,8 @@
       return CheckNullRefValue(cx, v, refval);
     case RefType::Eq:
       return CheckEqRefValue(cx, v, refval);
+    case RefType::I31:
+      return CheckI31RefValue(cx, v, refval);
     case RefType::Struct:
       return CheckStructRefValue(cx, v, refval);
     case RefType::Array:
@@ -177,22 +179,11 @@
 
 bool wasm::CheckAnyRefValue(JSContext* cx, HandleValue v,
                             MutableHandleAnyRef vp) {
-  if (v.isNull()) {
-    vp.set(AnyRef::null());
-    return true;
+  if (!AnyRef::fromJSValue(cx, v, vp)) {
+    MOZ_ASSERT(cx->isThrowingOutOfMemory());
+    return false;
   }
-
-  if (v.isObject()) {
-    JSObject& obj = v.toObject();
-    if (obj.is<WasmGcObject>()) {
-      vp.set(AnyRef::fromJSObject(obj.as<WasmGcObject>()));
-      return true;
-    }
-  }
-
-  JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr,
-                           JSMSG_WASM_BAD_ANYREF_VALUE);
-  return false;
+  return true;
 }
 
 bool wasm::CheckNullFuncRefValue(JSContext* cx, HandleValue v,
@@ -232,17 +223,14 @@
 
 bool wasm::CheckEqRefValue(JSContext* cx, HandleValue v,
                            MutableHandleAnyRef vp) {
-  if (v.isNull()) {
-    vp.set(AnyRef::null());
-    return true;
+  if (!AnyRef::fromJSValue(cx, v, vp)) {
+    MOZ_ASSERT(cx->isThrowingOutOfMemory());
+    return false;
   }
 
-  if (v.isObject()) {
-    JSObject& obj = v.toObject();
-    if (obj.is<WasmGcObject>()) {
-      vp.set(AnyRef::fromJSObject(obj.as<WasmGcObject>()));
-      return true;
-    }
+  if (vp.isNull() || vp.isI31() ||
+      (vp.isJSObject() && vp.toJSObject().is<WasmGcObject>())) {
+    return true;
   }
 
   JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr,
@@ -250,6 +238,22 @@
   return false;
 }
 
+bool wasm::CheckI31RefValue(JSContext* cx, HandleValue v,
+                            MutableHandleAnyRef vp) {
+  if (!AnyRef::fromJSValue(cx, v, vp)) {
+    MOZ_ASSERT(cx->isThrowingOutOfMemory());
+    return false;
+  }
+
+  if (vp.isNull() || vp.isI31()) {
+    return true;
+  }
+
+  JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr,
+                           JSMSG_WASM_BAD_I31REF_VALUE);
+  return false;
+}
+
 bool wasm::CheckStructRefValue(JSContext* cx, HandleValue v,
                                MutableHandleAnyRef vp) {
   if (v.isNull()) {
@@ -533,7 +537,35 @@
 template <typename Debug = NoDebug>
 bool ToWebAssemblyValue_eqref(JSContext* cx, HandleValue val, void** loc,
                               bool mustWrite64) {
-  return ToWebAssemblyValue_anyref(cx, val, loc, mustWrite64);
+  RootedAnyRef result(cx, AnyRef::null());
+  if (!CheckEqRefValue(cx, val, &result)) {
+    return false;
+  }
+  loc[0] = result.get().forCompiledCode();
+#ifndef JS_64BIT
+  if (mustWrite64) {
+    loc[1] = nullptr;
+  }
+#endif
+  Debug::print(*loc);
+  return true;
+}
+
+template <typename Debug = NoDebug>
+bool ToWebAssemblyValue_i31ref(JSContext* cx, HandleValue val, void** loc,
+                               bool mustWrite64) {
+  RootedAnyRef result(cx, AnyRef::null());
+  if (!CheckI31RefValue(cx, val, &result)) {
+    return false;
+  }
+  loc[0] = result.get().forCompiledCode();
+#ifndef JS_64BIT
+  if (mustWrite64) {
+    loc[1] = nullptr;
+  }
+#endif
+  Debug::print(*loc);
+  return true;
 }
 
 template <typename Debug = NoDebug>
@@ -659,6 +691,9 @@
         case RefType::Eq:
           return ToWebAssemblyValue_eqref<Debug>(cx, val, (void**)loc,
                                                  mustWrite64);
+        case RefType::I31:
+          return ToWebAssemblyValue_i31ref<Debug>(cx, val, (void**)loc,
+                                                  mustWrite64);
         case RefType::Struct:
           return ToWebAssemblyValue_structref<Debug>(cx, val, (void**)loc,
                                                      mustWrite64);